### ISSUE TITLE
Incomplete removal with specific symbolic links

### ISSUE THREAD

**Issue by git-bruh** [kiss-community/kiss#23]
test package:

```
#!/bin/sh -e

install -d "$1/tmp"

cd "$1/tmp"

mkdir 3.20

ln -s 3.20 3.21
```

The broken `3.21`﻿symlink remains when the package is uninstalled due to the logic in `pkg_remove_files`:

```
        # Remove files.
        if [ -f "$file" ] && [ ! -h "$file" ]; then
            rm -f "$file"

        # Remove file symlinks.
        elif [ -h "$file" ] && [ ! -d "$file" ]; then
            rm -f "$file"

        # Remove directories if empty.
        elif [ -d "$file" ] && [ ! -h "$file" ]; then
            rmdir "$file" 2>/dev/null ||:
        fi
```

The condition `[ -h ] && [ ! -d ]` will eval to true only when the `3.21` directory is removed. Here the removal breaks because the check happens before the `3.21` directory is removed. File symlinks aren't affected since they're always deleted.

This can ofcourse be solved by removing the `[ ! -d ]` check, I'm not sure why this check is needed anyways, since directory symlinks can be removed by a simple `rm` too.

**Comment by aaronNGi** [comment]
Since `pkg_remove_files()` will try to remove everything from a packages manifest (which includes top level dirs), it could potentially remove symlinks like /bin, /lib, etc. So I guess Dylans work-around was to simply never remove symlinks pointing to a directory.

**Comment by cemkeylan** [comment]
Yes, that's the exact reason behind it. See kisslinux/kiss#131

**Comment by dilyn-corner** [comment]
Theoretically we could get rid of this if we just... forced `kiss` to ignore every directory that should be persistent. But that's gross, and I don't think there's a not-gross way to do it. 
Is this an actual problem, or just a theoretical one?

**Comment by git-bruh** [comment]
I encountered this with some old GTK theme, but it's probably specific to that one since i haven't seen any other theme create such links.

I guess this can be deemed as a non-issue as there's next to no packages using symlinks to a directory, and this problem doesn't exist with them.

```
-> find /usr -type l -exec test -d {} \; -exec ls -l {} \;
lrwxrwxrwx    1 root     root            27 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/mips -> ../../../arch/mips/boot/dts
lrwxrwxrwx    1 root     root            25 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/sh -> ../../../arch/sh/boot/dts
lrwxrwxrwx    1 root     root            28 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/nios2 -> ../../../arch/nios2/boot/dts
lrwxrwxrwx    1 root     root            31 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/openrisc -> ../../../arch/openrisc/boot/dts
lrwxrwxrwx    1 root     root            29 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/xtensa -> ../../../arch/xtensa/boot/dts
lrwxrwxrwx    1 root     root            28 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/dt-bindings -> ../../../include/dt-bindings
lrwxrwxrwx    1 root     root            28 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/arm64 -> ../../../arch/arm64/boot/dts
lrwxrwxrwx    1 root     root            30 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/powerpc -> ../../../arch/powerpc/boot/dts
lrwxrwxrwx    1 root     root            28 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/h8300 -> ../../../arch/h8300/boot/dts
lrwxrwxrwx    1 root     root            26 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/arm -> ../../../arch/arm/boot/dts
lrwxrwxrwx    1 root     root            26 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/arc -> ../../../arch/arc/boot/dts
lrwxrwxrwx    1 root     root            33 Apr 29 16:09 /usr/src/linux-5.12/scripts/dtc/include-prefixes/microblaze -> ../../../arch/microblaze/boot/dts
lrwxrwxrwx    1 root     root             6 Apr 28 18:52 /usr/local/share/man -> ../man
lrwxrwxrwx    1 root     root             3 Apr 28 18:52 /usr/lib64 -> lib
lrwxrwxrwx    1 root     root             3 Apr 28 18:52 /usr/sbin -> bin
lrwxrwxrwx    1 root     root             9 Apr 29 20:41 /usr/share/themes/gruvbox-dark-gtk/gtk-3.20/assets -> ../assets
lrwxrwxrwx    1 root     root            19 Apr 29 16:11 /usr/lib/modules/5.12.0/source -> /usr/src/linux-5.12
lrwxrwxrwx    1 root     root            19 Apr 29 16:11 /usr/lib/modules/5.12.0/build -> /usr/src/linux-5.12
lrwxrwxrwx    1 root     root            17 Apr 29 14:34 /usr/lib/terminfo -> ../share/terminfo
```

Maybe there could just be a warning like for files in `/etc` since most proper packages just have stuff in /usr/{bin,lib} and not symlinks like sbin ?�

**Comment by aaronNGi** [comment]
Instead of adding code which guards certain files and dirs, we could add a check, which protects top-level symlinks. Only slightly more elegant though. I think the best would be to test if the linked directory still exists but then the order of removal matters, meaning we can't really do that when the symlink would get deleted before the linked directory.

So if this is not a huge issue, keeping it as is, is the best solution on my book. A warning would be nice but that would also require special rules or otherwise it will warn about not deleting the /bin symlink on most package removals.

**Comment by dilyn-corner** [comment]
This should be resolved now with 754c9424f1b74af494b96fcb2dab7d9f51f14557
